File: ecmascript/compiler/codegen/maple/maple_ir/src/lexer.cpp
/*
 * Copyright (c) 2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "lexer.h"
#include <cmath>
#include <climits>
#include <cstdlib>
#include "mpl_logging.h"
#include "debug_info.h"
#include "mir_module.h"
#include "securec.h"
#include "utils.h"

namespace maple {
int32 HexCharToDigit(char c)
{
    int32 ret = utils::ToDigit<16, int32>(c);
    return (ret != INT32_MAX ? ret : 0);
}

// Read (next) line from the MIR (text) file, and return the read
// number of chars.
// if the line is empty (nothing but a newline), returns 0.
// if EOF, return -1.
// The trailing new-line character has been removed.
int MIRLexer::ReadALine()
{
    if (airFile == nullptr) {
        line = "";
        return -1;
    }

    curIdx = 0;
    if (!std::getline(*airFile, line)) {  // EOF
        line = "";
        airFile = nullptr;
        currentLineSize = 0;
        return -1;
    }

    RemoveReturnInline(line);
    currentLineSize = line.length();
    return currentLineSize;
}

int MIRLexer::ReadALineByMirQueue()
{
    if (mirQueue.empty()) {
        line = "";
        return -1;
    }
    curIdx = 0;
    line = mirQueue.front();
    RemoveReturnInline(line);
    currentLineSize = line.length();
    mirQueue.pop();
    return currentLineSize;
}

MIRLexer::MIRLexer(MIRModule &mod)
    : module(mod), seenComments(mod.GetMPAllocator().Adapter()), keywordMap(mod.GetMPAllocator().Adapter())
{
    // initialize keywordMap
    keywordMap.clear();
#define KEYWORD(STR)                \
    {                               \
        std::string str;            \
        str = #STR;                 \
        keywordMap[str] = TK_##STR; \
    }
#include "keywords.def"
#undef KEYWORD
}

void MIRLexer::PrepareForFile(const std::string &filename)
{
    // open MIR file
    airFileInternal.open(filename);
    CHECK_FATAL(airFileInternal.is_open(), "cannot open MIR file %s\n", &filename);

    airFile = &airFileInternal;
    // try to read the first line
    if (ReadALine() < 0) {
        lineNum = 0;
    } else {
        lineNum = 1;
    }
    module.GetDbgInfo()->UpdateMsg(lineNum, line.c_str());
    kind = TK_invalid;
}

void MIRLexer::PrepareForString(const std::string &src)
{
    SetMirQueue(src);
    if (ReadALineByMirQueue() < 0) {
        lineNum = 0;
    } else {
        lineNum = 1;
    }
    module.GetDbgInfo()->UpdateMsg(lineNum, line.c_str());
    kind = TK_invalid;
}

void MIRLexer::GenName()
{
    uint32 startIdx = curIdx;
    char c = GetNextCurrentCharWithUpperCheck();
    CHECK_FATAL(curIdx > 0, "must not be zero");
    char cp = GetCharAt(curIdx - 1);
    if (c == '@' && (cp == 'h' || cp == 'f')) {
        // special pattern for exception handling labels: catch or finally
        c = GetNextCurrentCharWithUpperCheck();
    }
    while (utils::IsAlnum(c) || c < 0 || c == '_' || c == '$' || c == ';' || c == '/' || c == '|' || c == '.' ||
           c == '?' || c == '@') {
        c = GetNextCurrentCharWithUpperCheck();
    }
    name = line.substr(startIdx, curIdx - startIdx);
}

TokenKind MIRLexer::GetSpecialFloatConst()
{
    constexpr uint32 lenSpecFloat = 4;
    constexpr uint32 lenSpecDouble = 3;
    if (line.compare(curIdx, lenSpecFloat, "inff") == 0 &&
        !utils::IsAlnum(GetCharAtWithUpperCheck(curIdx + lenSpecFloat))) {
        curIdx += lenSpecFloat;
        theFloatVal = -INFINITY;
        return TK_floatconst;
    }
    if (line.compare(curIdx, lenSpecDouble, "inf") == 0 &&
        !utils::IsAlnum(GetCharAtWithUpperCheck(curIdx + lenSpecDouble))) {
        curIdx += lenSpecDouble;
        theDoubleVal = -INFINITY;
        return TK_doubleconst;
    }
    if (line.compare(curIdx, lenSpecFloat, "nanf") == 0 &&
        !utils::IsAlnum(GetCharAtWithUpperCheck(curIdx + lenSpecFloat))) {
        curIdx += lenSpecFloat;
        theFloatVal = -NAN;
        return TK_floatconst;
    }
    if (line.compare(curIdx, lenSpecDouble, "nan") == 0 &&
        !utils::IsAlnum(GetCharAtWithUpperCheck(curIdx + lenSpecDouble))) {
        curIdx += lenSpecDouble;
        theDoubleVal = -NAN;
        return TK_doubleconst;
    }
    return TK_invalid;
}

TokenKind MIRLexer::GetHexConst(uint32 valStart, bool negative)
{
    char c = GetCharAtWithUpperCheck(curIdx);
    if (!isxdigit(c)) {
        name = line.substr(valStart, curIdx - valStart);
        return TK_invalid;
    }
    uint64 tmp = static_cast<uint32>(HexCharToDigit(c));
    c = GetNextCurrentCharWithUpperCheck();
    while (isxdigit(c)) {
        tmp = (tmp << k16BitShift) + static_cast<uint32>(HexCharToDigit(c));
        c = GetNextCurrentCharWithUpperCheck();
    }
    theIntVal = static_cast<uint64>(static_cast<uint64>(tmp));
    if (negative) {
        theIntVal = -theIntVal;
    }
    theFloatVal = static_cast<float>(theIntVal);
    theDoubleVal = static_cast<double>(theIntVal);
    if (negative && theIntVal == 0) {
        theFloatVal = -theFloatVal;
        theDoubleVal = -theDoubleVal;
    }
    name = line.substr(valStart, curIdx - valStart);
    return TK_intconst;
}

TokenKind MIRLexer::GetIntConst(uint32 valStart, bool negative)
{
    auto negOrSelf = [negative](uint64 val) { return negative ? ~val + 1 : val; };

    theIntVal = static_cast<uint64>(HexCharToDigit(GetCharAtWithUpperCheck(curIdx)));

    uint64 radix = theIntVal == 0 ? 8 : 10;

    char c = GetNextCurrentCharWithUpperCheck();

    for (theIntVal = negOrSelf(theIntVal); isdigit(c); c = GetNextCurrentCharWithUpperCheck()) {
        theIntVal = (theIntVal * radix) + negOrSelf(HexCharToDigit(c));
    }

    if (c == 'u' || c == 'U') {  // skip 'u' or 'U'
        c = GetNextCurrentCharWithUpperCheck();
        if (c == 'l' || c == 'L') {
            c = GetNextCurrentCharWithUpperCheck();
        }
    }

    if (c == 'l' || c == 'L') {
        c = GetNextCurrentCharWithUpperCheck();
        if (c == 'l' || c == 'L' || c == 'u' || c == 'U') {
            ++curIdx;
        }
    }

    name = line.substr(valStart, curIdx - valStart);

    if (negative) {
        theFloatVal = static_cast<float>(static_cast<int64>(theIntVal));
        theDoubleVal = static_cast<double>(static_cast<int64>(theIntVal));

        if (theIntVal == 0) {
            theFloatVal = -theFloatVal;
            theDoubleVal = -theDoubleVal;
        }
    } else {
        theFloatVal = static_cast<float>(theIntVal);
        theDoubleVal = static_cast<double>(theIntVal);
    }

    return TK_intconst;
}

TokenKind MIRLexer::GetTokenWithPrefixDollar()
{
    // token with prefix '$'
    char c = GetCharAtWithUpperCheck(curIdx);
    if (utils::IsAlpha(c) || c == '_' || c == '$') {
        GenName();
        return TK_gname;
    } else {
        // for error reporting.
        const uint32 printLength = 2;
        name = line.substr(curIdx - 1, printLength);
        return TK_invalid;
    }
}

TokenKind MIRLexer::GetTokenWithPrefixPercent()
{
    // token with prefix '%'
    char c = GetCharAtWithUpperCheck(curIdx);
    if (isdigit(c)) {
        int valStart = static_cast<int>(curIdx) - 1;
        theIntVal = static_cast<uint64>(HexCharToDigit(c));
        c = GetNextCurrentCharWithUpperCheck();
        while (isdigit(c)) {
            theIntVal = (theIntVal * 10) + static_cast<uint64>(HexCharToDigit(c)); // 10 for decimal
            DEBUG_ASSERT(theIntVal >= 0, "int value overflow");
            c = GetNextCurrentCharWithUpperCheck();
        }
        name = line.substr(valStart, curIdx - valStart);
        return TK_preg;
    }
    if (utils::IsAlpha(c) || c == '_' || c == '$') {
        GenName();
        return TK_lname;
    }
    if (c == '%' && utils::IsAlpha(GetCharAtWithUpperCheck(curIdx + 1))) {
        ++curIdx;
        GenName();
        return TK_specialreg;
    }
    return TK_invalid;
}

TokenKind MIRLexer::GetTokenWithPrefixAmpersand()
{
    // token with prefix '&'
    char c = GetCurrentCharWithUpperCheck();
    if (utils::IsAlpha(c) || c == '_') {
        GenName();
        return TK_fname;
    }
    // for error reporting.
    constexpr uint32 printLength = 2;
    CHECK_FATAL(curIdx > 0, "must not be zero");
    name = line.substr(curIdx - 1, printLength);
    return TK_invalid;
}

TokenKind MIRLexer::GetTokenWithPrefixAtOrCircumflex(char prefix)
{
    // token with prefix '@' or `^`
    char c = GetCurrentCharWithUpperCheck();
    if (utils::IsAlnum(c) || c < 0 || c == '_' || c == '@' || c == '$' || c == '|') {
        GenName();
        if (prefix == '@') {
            return TK_label;
        }
        return TK_prntfield;
    }
    return TK_invalid;
}

TokenKind MIRLexer::GetTokenWithPrefixExclamation()
{
    // token with prefix '!'
    char c = GetCurrentCharWithUpperCheck();
    if (utils::IsAlpha(c)) {
        GenName();
        return TK_typeparam;
    }
    // for error reporting.
    const uint32 printLength = 2;
    CHECK_FATAL(curIdx > 0, "must not be zero");
    name = line.substr(curIdx - 1, printLength);
    return TK_invalid;
}

TokenKind MIRLexer::GetTokenWithPrefixQuotation()
{
    if (GetCharAtWithUpperCheck(curIdx + 1) == '\'') {
        theIntVal = GetCharAtWithUpperCheck(curIdx);
        constexpr uint32 hexLength = 2;
        curIdx += hexLength;
        return TK_intconst;
    }
    return TK_invalid;
}

TokenKind MIRLexer::GetTokenSpecial()
{
    --curIdx;
    char c = GetCharAtWithLowerCheck(curIdx);
    if (utils::IsAlpha(c) || c < 0 || c == '_') {
        GenName();
        TokenKind tk = keywordMap[name];
        switch (tk) {
            case TK_nanf:
                theFloatVal = NAN;
                return TK_floatconst;
            case TK_nan:
                theDoubleVal = NAN;
                return TK_doubleconst;
            case TK_inff:
                theFloatVal = INFINITY;
                return TK_floatconst;
            case TK_inf:
                theDoubleVal = INFINITY;
                return TK_doubleconst;
            default:
                return tk;
        }
    }
    MIR_ERROR("error in input file\n");
    return TK_eof;
}

TokenKind MIRLexer::NextToken()
{
    return kind;
}
}  // namespace maple

File: ecmascript/arksteed/test/weekly_preheat/basic/lexer.ts
/*
 * Copyright (c) 2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// 词/token，词法分析结果
class Token {
  kind: string;

  string: string;

  sourceLineNumber?: number;

  userLineNumber?: number;

  value: number | string | null = null;

  constructor(kind: string, string: string, sourceLineNumber?: number, userLineNumber?: number) {
    this.kind = kind;
    this.string = string;
    this.sourceLineNumber = sourceLineNumber;
    this.userLineNumber = userLineNumber;
  }
}

/// 词法分析器
class Lexer implements Iterable<Token>, Iterator<Token> {
  /// 源代码
  source: string;
  /// 源代码行号
  sourceLineNumber: number = 1;
  /// 是否为行首
  isLineStart: boolean;
  /// 用户输入的行号
  userLineNumber?: number;
  /// 当前解析的代码，也记录了解析的进度
  currentCode: string | null = null;
  /// 所有的代码行
  lines: string[];
  /// 标识符正则
  isIdentifier: RegExp = new RegExp('^[a-z_]([a-z0-9_]*)');
  /// 数字正则
  isNumber: RegExp = new RegExp('^(([0-9]+(\\.([0-9]*))?)|(\\.[0-9]+)(e([+-]?)([0-9]+))?)');

  /// 字符串正则
  isString: RegExp = new RegExp('^\\"([^\\"]|(\\"\\"))*\\"');

  /// 关键字正则
  isKeyword: RegExp = new RegExp(
    '^((base)|(data)|(def)|(dim)|(end)|(for)|(go)|(gosub)|(goto)|(if)|(input)|(let)|(next)|(on)|(option)|(print)|(randomize)|(read)|(restore)|(return)|(step)|(stop)|(sub)|(then)|(to))'
  );
  /// 操作符正则
  isOperator: RegExp = new RegExp('^(-|\\+|\\*|\\/|\\^|\\(|\\)|(<[>=]?)|(>=?)|=|,|\\$|;)');

  isRem: RegExp = new RegExp('^rem\\s.*');

  constructor(source: string) {
    this.source = source;
    this.lines = source.split('\n');
    this.currentCode = this.lines[this.sourceLineNumber - 1];
    this.isLineStart = true;
  }

  /// 匹配并去除空格
  consumeWhitespace(): void {
    let reg = new RegExp('^\\s+');
    let matches = this.currentCode!.match(reg);
    if (matches !== null && matches.length > 0) {
      this.currentCode = this.currentCode!.slice(matches[0].length);
    }
  }

  /// 迭代器 词法解析 provide the next token
  next(): IteratorResult<Token> {
    if (this.currentCode === null) {
      return { done: true, value: null };
    }
    if (this.isLineStart) {
      this.consumeWhitespace();
      let isnum = new RegExp('^[0-9]+');
      let matches = this.currentCode.match(isnum);
      if (matches !== null && matches.length > 0) {
        this.userLineNumber = Math.floor(Number(matches[0]));
        this.currentCode = this.currentCode.slice(matches[0].length);
        this.isLineStart = false;
        return {
          value: new Token('userLineNumber', matches[0], this.sourceLineNumber, this.userLineNumber!),
          done: false
        };
      }
    } else {
      this.consumeWhitespace();
      if (this.currentCode!.length === 0) {
        this.isLineStart = true;
        if (this.sourceLineNumber === this.lines.length) {
          this.currentCode = null;
          return {
            done: false,
            value: new Token('newLine', '\n', this.sourceLineNumber, this.userLineNumber)
          };
        } else {
          let curLineNumber = this.sourceLineNumber;
          this.sourceLineNumber += 1;
          this.currentCode = this.lines[this.sourceLineNumber - 1];
          return {
            done: false,
            value: new Token('newLine', '\n', curLineNumber, this.userLineNumber!)
          };
        }
      }
      return this.parseToken();
    }
    return { value: null, done: true };
  }

  parseToken(): IteratorResult<Token> {
    let matches = this.currentCode!.match(this.isKeyword);
    if (matches !== null && matches.length > 0) {
      this.currentCode = this.currentCode!.slice(matches[0].length);
      return { done: false, value: new Token('keyword', matches[0], this.sourceLineNumber, this.userLineNumber) };
    }
    matches = this.currentCode!.match(this.isIdentifier);
    if (matches !== null && matches.length > 0) {
      this.currentCode = this.currentCode!.slice(matches[0].length);
      return {
        done: false,
        value: new Token('identifier', matches[0], this.sourceLineNumber, this.userLineNumber!)
      };
    }
    matches = this.currentCode!.match(this.isNumber);
    if (matches !== null && matches.length > 0) {
      this.currentCode = this.currentCode!.slice(matches[0].length);
      let token = new Token('number', matches[0], this.sourceLineNumber, this.userLineNumber!);
      token.value = Number(matches[0]);
      return { done: false, value: token };
    }
    matches = this.currentCode!.match(this.isString);
    if (matches !== null && matches.length > 0) {
      this.currentCode = this.currentCode!.slice(matches[0].length);
      let str = matches[0];
      str = str.slice(1, str.length - 1);
      let strToken = new Token('string', str, this.sourceLineNumber, this.userLineNumber!);
      strToken.value = str;
      return { done: false, value: strToken };
    }
    matches = this.currentCode!.match(this.isOperator);
    if (matches !== null && matches.length > 0) {
      this.currentCode = this.currentCode!.slice(matches[0].length);
      return {
        done: false,
        value: new Token('operator', matches[0], this.sourceLineNumber, this.userLineNumber!)
      };
    }
    matches = this.currentCode!.match(this.isRem);
    if (matches !== null && matches.length > 0) {
      this.currentCode = this.currentCode!.slice(matches[0].length);
      return {
        done: false,
        value: new Token('remark', matches[0], this.sourceLineNumber, this.userLineNumber!)
      };
    }
    return { done: false, value: new Token('Error', 'Error', -1, -1) };
  }

  [Symbol.iterator](): Iterator<Token> {
    return this;
  }
}

export { Lexer, Token };
